Add distributed optimizer section to distributed autograd design doc.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30069 Polish rpc docstring.
* **#30068 Add distributed optimizer section to distributed autograd design doc.**

Differential Revision: [D18556536](https://our.internmc.facebook.com/intern/diff/D18556536/)

Commit:
Add distributed optimizer section to distributed autograd design doc.



Differential Revision: [D18556536](https://our.internmc.facebook.com/intern/diff/D18556536/)

[ghstack-poisoned]

mrshenli commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -1,3 +1,6 @@
+.. warning::
+  The :ref:`distributed-rpc-framework` is experimental and subject to change.
```
Since this is distributed_autograd doc, should it actually say **distributed autograd** is experimental?

pritamdamania87 replied:
This actually links to the `Distributed RPC Framework` doc page and distributed autograd is a subsection there.

mrshenli commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -1,3 +1,6 @@
+
 .. _distributed-autograd-design:
 
 .. warning::
```
this is duplicated with the above warning block.

mrshenli commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -295,6 +298,84 @@ you can refer to **Distributed Autograd Algorithm Smart mode** section in the
+   optimize. These could also be local parameters wrapped within a local
+   ``RRef``.
+2. Takes a :class:`~torch.optim.Optimizer` class as the local
+   optimizer to run on all nodes.
```
all nodes -> all distinct RRef owners

mrshenli commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -295,6 +298,84 @@ you can refer to **Distributed Autograd Algorithm Smart mode** section in the
+   ``RRef``.
+2. Takes a :class:`~torch.optim.Optimizer` class as the local
+   optimizer to run on all nodes.
+3. The distributed optimizer creates an instance of the local ``Optimizer`` on
```
is 3 kind of duplicated with 2?

pritamdamania87 replied:
2 is mentioning the inputs to the distributed optimizer, 3 is just mentioning that we use those to create RRefs remotely.

mrshenli commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -295,6 +298,84 @@ you can refer to **Distributed Autograd Algorithm Smart mode** section in the
+   each of the worker nodes and holds an ``RRef`` to them.
+4. When :meth:`torch.distributed.optim.DistributedOptimizer.step` is invoked,
+   the distributed optimizer uses RPC to remotely execute all the local
+   optimizers on the appropriate remote workers.
```
Do we need to mentioned the optimizer param/grad sync/lock behavior here?

mrshenli commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -295,6 +298,84 @@ you can refer to **Distributed Autograd Algorithm Smart mode** section in the
+  from torch import optim
+  from torch.distributed.optim import DistributedOptimizer
+  from tempfile import NamedTemporaryFile
+  import multiprocessing as mp
```
Let's sort this block

rohan-varma commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -295,6 +298,84 @@ you can refer to **Distributed Autograd Algorithm Smart mode** section in the
+  # Run two workers.
+  for i in range(2):
+      p = mp.Process(target=run_process, args=(i, (i + 1) % 2, file_name))
+      p.start()
```
nit: `processes.append(p)`

pritamdamania87 replied:
Good catch, not sure how I missed that :)

rohan-varma commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -295,6 +298,84 @@ you can refer to **Distributed Autograd Algorithm Smart mode** section in the
+
+  def run_process(self_rank, dst_rank, file_name):
+      _run_process(self_rank, dst_rank, file_name)
+      rpc.join_rpc()
```
I'm assuming we have to wrap the above function due to rrefs leaking issue?

pritamdamania87 replied:
Yes

Commit:
Update on "Add distributed optimizer section to distributed autograd design doc."


Differential Revision: [D18556536](https://our.internmc.facebook.com/intern/diff/D18556536/)

[ghstack-poisoned]

mrshenli commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -295,6 +295,87 @@ you can refer to **Distributed Autograd Algorithm Smart mode** section in the
+5. If multiple concurrent distributed optimizers are updating the same
+   parameters on a worker, these updates are serialized via a lock.
+
+Simple end to end example
```
Other section names use first capital letter

pritamdamania87 replied:
This is true for top level sections but not necessarily other sections. I mostly capitalized based on whether it looked appropriate :)

mrshenli reviewed (approved):
LGTM

rohan-varma reviewed (approved):
Thanks, looks good!
